CC: CMake Tutorial and Reference Resources (#1259) (#1276)
betterscientificsoftware/bssw.io: CuratedContent/CMakeTutorialAndReferenceResources.md

# CMake Tutorial and Reference Resources

<!--deck text start-->

> The CMake meta-build system is a large and powerful set of tools to create cross-platform builds from source for C, C++, Fortran and many other compiled programming languages.

Author: Okay.  I mention "CMake meta-build system" in the longer introduction paragraph.

> The CMake meta-build system is a large and powerful set of tools to create cross-platform builds from source for C, C++, Fortran and many other compiled programming languages.

Author: Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

> CMake continues to rapidly evolve and best practices can change from year to year (and even from release to release in some cases).

Reviewer: ```suggestion
CMake continues to rapidly evolve and best practices in its use are changing almost as rapidly.
```

> Here we provide links and descriptions for the best tutorial and reference materials that exist for CMake to help stay on the forefront of effective CMake usage.

Reviewer: ```suggestion
Here we provide links and descriptions for the best tutorial and reference materials that exist for CMake to help stay current in effective CMake usage.
```

<!--deck text end-->

#### Contributed by [Roscoe A. Bartlett](https://github.com/bartlettroscoe)
#### Publication date: ???

Resource information | Details 
:--- | :--- 
Book title | [Professional CMake](https://crascit.com/professional-cmake/)
Authors | Craig Scott
Publication | Self published, new editions put out after each new CMake release
Web links | https://crascit.com/professional-cmake/

Resource information | Details 
:--- | :--- 
Web Page | [CMake Reference Documentation](https://cmake.org/cmake/help/latest/)
Authors | Kitware Staff
Web links | https://cmake.org/cmake/help/latest/ (and drop-down section of all CMake versions)

Resource information | Details 
:--- | :--- 
Book title | [Mastering CMake](https://crascit.com/professional-cmake/)
Authors | Kitware Staff
Publication | Year 2022, Self published, currently not versioned
Web links | https://crascit.com/professional-cmake/

> The CMake meta-build tool and ecosystem (which includes the tools CTest and CPack with each CMake installation) continues to evolve and improve fairly rapidly from release to release.

Reviewer: Does this expository text apply to the one item *just above*, to all of the items above or to the item coming next? Greater verticle spacing or a horizontal linke (in GFM thats a sep. line with just `---`) might help to make this clear

Author: All of the above.  So you suggest breaking out the intro with  `---`?  I will try that and see how it looks.

Author: Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

Author: Addressed in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

> Kitware (the company that owns the primary CMake repository and provides the official releases of CMake) puts out three major releases of CMake each year (or about every four months).

Reviewer: ```suggestion
Kitware (the organization responsible for CMake development) makes three major releases each year (or about every four months).
```

> The best practices for using CMake evolve as new versions of CMake are released.

Reviewer: ```suggestion
Best practices for using CMake evolve as new versions are released.
```

> (For example, at the time of this writing, CMake support for a feature called Generator Expressions is still rapidly evolving and can have a dramatic impact on the best way to write CMakeList.txt files.)

Reviewer: ```suggestion
For example, at the time of this writing, CMake support for [Generator Expressions](https://cmake.org/cmake/help/latest/manual/cmake-generator-expressions.7.html#manual:cmake-generator-expressions(7)) is rapidly evolving and can have a dramatic impact on the best way to write `CMakeList.txt` files.)
```

> Therefore, unlike other programming tools, trying to use web searches to find solutions for CMake can be ineffective; because even advice from just a year ago on Stack Overflow or Stack Exchange can be outdated.

Reviewer: Not sure I follow. Can't someone just be sure to to use a search that restricts results to nothing less than 6 months or a year old?

Author: I would not say that.  Some CMake problems like how to handle semi-colons ';' chars in the test property ENVIRONMENT are the same now as the were 10+ years ago (and Kitware still does not have a solution to that problem or even document how to successfully handle semi-colons ';' on these properties, even through the workarounds are the same has they have been for many years).  However, advice on how to best generate files in many cases (i.e. using `file(GENERATE ...)` instead of `configure_file(...)`) changed dramatically a few years ago.  And as more commands support generator expressions, the suggested way to solve certain problems will be impacted.

> Therefore, this page gives advice based on 13+ years of CMake usage and hundreds of hours of searching for and reading CMake documentation, especially in the years 2021-2022.

Reviewer: What is `this page` referring to here? Can you provide link. Is it the stuff above? Oh, wait, I think you mean this CC article.

```suggestion
The bssw.io editorial team has created this article informed by 13+ years of [advanced CMake usage](https://tribits.org/doc/TribitsDevelopersGuide.html) and hundreds of hours of searching for and reading CMake documentation, especially in the years 2021-2022.
```

Author: > What is `this page` referring to here?

I mean, "the article you are reading".

<img src='../images/ProfessionalCMake.jpg' class='page'/>

Arguably, the best resource for finding up-to-date guidance on how to use CMake is the most recent edition of the book ***Professional CMake*** (the 11th edition covering CMake versions up to 3.22 as of this writing).

> The 560+ page 11th edition of the book contains excellent tutorial material for CMake and yet goes into deep concepts in the CMake language where needed (such as the gotchas and workarounds for dealing with lists and function/macro arguments with data containing quotes `"`, spaces ` ` and semi-colons `;` which is one of the most confusing aspects to using CMake).

Reviewer: ```suggestion
The book contains excellent tutorial material for CMake and yet goes into deep concepts in the CMake language where needed (such as the gotchas and workarounds for dealing with lists and function/macro arguments with data containing quotes `"`, spaces ` ` and semi-colons `;` a frequently misunderstood aspect of CMake).
```

This book covers every basic CMake feature in great detail.
It explains modern CMake idioms and best practices.

> Also, it covers older CMake commands and idioms to avoid and why.

Reviewer: ```suggestion
Also, it covers older CMake commands and idioms to avoid and explains why.
```

> This book seems to answer the "why" and big picture questions for CMake in a way that no other material has matched.

Reviewer: ```suggestion
This book addresses the "why" and big picture questions about CMake in a way that no other material has matched.
```

> It also provides really critical low-level details and tips that are absent in even the official CMake reference documentation.

Reviewer: ```suggestion
It also provides really critical low-level details and tips that are absent in even the [official CMake reference documentation](https://cmake.org/cmake/help/latest/index.html).
```

This book is self published by the author in a downloadable PDF format with new editions coming out after each new CMake release (i.e. three new editions each year).
The current edition of the book costs $30 which includes 9 updates for new editions of the book as they come out.
(Therefore, $30 would cover every new edition over a 3 year period which includes 9 new CMake releases or 6 years if only downloading every other updated edition across 18 CMake releases.)

> The author is not currently a Kitware employee but is a significant contributor to CMake and collaborates closely with Kitware staff.

Reviewer: ```suggestion
The author is a significant contributor to CMake and collaborates closely with Kitware staff.
```

> The author starts evaluating and updating the book at the beginning of each new CMake release cycle (i.e. right after the first release candidate `vX.Y.0-rc1` is tagged).

Reviewer: Not sure this and following sentences are needed. Makes for wordiness but not much in the way of truly needed information.

Author: Fair enough.  The CC article already says that a new edition is put out after each CMake release.  The process by which new editions of the book are created in relation to CMake releases is not really something our readers need to know.  (The reason I mentioned this was to highlight how embedded Craig Scott is with CMake development and releases.)

Author: Addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

Through this process, the author often provides feedback to Kitware and help addressing issues with new CMake features and changes to existing CMake functionality which feeds back into bug fixes and better CMake reference and release notes documentation.

Another important source of CMake documentation is Kitware's official online ***CMake Reference Documentation***.
This documentation is produced and versioned for every CMake release and contains a wealth of information including:

> * Reference material for the entire CMake language including for every supported command, property, variable and related topic

Reviewer: Here you refer to the CMake *language* and first line refer to CMake *meta-build system*. Might wanna stick with just one of these. I think *language is fine.

Author: Technically speaking, there is more to CMake than just the CMake language.  There are the executables themselves and their command-line arguments.  There are input files and input file formats that feed into CMake (e.g. JSON formats for certain input files like for presets and test resources).    There are standard CMake modules and their documentation.  And there are many other topics that all fall under "CMake" that are not the CMake language itself.

Reviewer: Right but was there a specific reason you used `CMake language` here but used `CMake meta-build` somewhere else (and I think I recall running across `CMake ecosystem` as well)?

Whatever you want to call it (a language, a meta-build system, an ecosystem) is fine. But, I think it should be consistent. And, for what its worth, I think its fine to probably just call it solely by name, `CMake` as in...

- Reference material for CMake including...
- CMake is a large and powerful set of tools...
- ...is still the best advice for the current state of CMake

That avoids entirely any potential cognitive dissidence for what any reader might call it.

Author: I think I can see your point.  However, these adjectives are meant to convey extra meaning.  Here are the definitions:

* Meta-build system: A tool that takes a high-level platform-independent description of a set of software to build and install and then generates the concrete build system (e.g. using Make or Ninja) to actually build the software for a specific system or machine.
* Ecosystem: A collection of related but (in some way) independent software packages or entities.  (In this case CMake, CTest, CPack, CDash and perhaps some related efforts make up the "CMake ecosystem".)
* Language: A set of commands and constructs to execute operations and logic in a computer program.

Are these not well-known definitions?

Author: > Meta-build system

That is actually not consistent with the definition of "meta-build" tool given at:

* https://en.wikipedia.org/wiki/List_of_build_automation_software

But it is consistent with the definition of "meta-build system" which includes CMake at:

* https://sinban04.github.io/build-system/
* http://www.kumarakshay.me/cmake/
* https://alexreinking.com/blog/cmake-is-a-build-system.html

So I think we should remove the term "meta-build system" because there does not seem to be agreement on the definition of a "meta-build system".

Author: Hopefully addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/5fe02d6d33b92c7d64e2a1a9b95d1a18c8f40feb

* Documentation for the large collection of standard `*.cmake` modules that are installed along with CMake
* Targeted guides including:
  * CMake Tutorial (for developers contributing to projects that use CMake)
  * User interaction Guide (for users trying to configure, build, install and test projects that use CMake)
* Release notes for each version of CMake
* Command-line documentation for the `cmake`, `ctest`, and `cpack` executables

> As a developer contributing to a project using CMake (i.e. editing the project's `CMakeLists.txt` files), when consulting the official reference documentation, it is advisable to only view documentation for the minimum version of CMake supported by the project (i.e. for the version `X.Y` listed in the project's top-level `CMakeLists.txt` file calling `cmake_minimum_required(VERSION X.Y)`).

Reviewer: ```suggestion
As a developer contributing to a project using CMake (i.e. editing the project's `CMakeLists.txt` files), when consulting the official reference documentation, it is advisable to view documentation for only the minimum version of CMake supported by the project (i.e. for the version `X.Y` listed in the project's top-level `CMakeLists.txt` file calling `cmake_minimum_required(VERSION X.Y)`).
```

> For example, if minimum version was 3.17, then one would select that version in the drop-down box (taking you to https://cmake.org/cmake/help/v3.17/).  This helps to avoid accidentally using CMake features not present in the minimum CMake version supported by the project.

Reviewer: ```suggestion
For example, if minimum version is 3.17, then one would select that version in the drop-down box (taking you to https://cmake.org/cmake/help/v3.17/).  This helps to avoid accidentally using CMake features available only in newer versions.
```

If one does not want to spend the $30 for the book "Professional CMake", then one may consider Kitware's ***Mastering CMake*** book.
This is an updated, free online version of Kitware's classic "Mastering CMake" paperback book that used to be sold, (last updated for [CMake 3.1 published in 2015](https://www.amazon.com/Mastering-CMake-Ken-Martin/dp/1930934319)).
It also contains tutorial coverage of key CMake concepts and other basic tutorial content (but usually at a less complete and detailed level compared to "Professional CMake").

> Note that the "guides" provided in the online CMake Reference Documentation are also exactly duplicated in the online "Mastering CMake" book.

Reviewer: ```suggestion
Note that the "guides" provided in the online CMake Reference Documentation are also precisely duplicated in the online "Mastering CMake" book.
```

> Finally, if one searches the above mentioned documentation sources "Professional CMake", the online CMake Reference Documentation, and "Mastering CMake" and does not find the answer, then one can turn to web searches.

Reviewer: I would remove this sentence and the next two sentences.

I might add a note that when search on-line sources for CMake documentation, its a good idea to date-restrict the results to material less than one year old.

Author: As I mentioned [above](https://github.com/betterscientificsoftware/bssw.io/pull/1276#discussion_r831594517), you don't want to restrict your web searches to just recent hits but you also want to be careful to make sure this is the best advice given the current state of CMake.

The problem is that the internet contains a lot of material and advice on how to use CMake that is badly out of date given the current state of modern CMake.  But there are also long-standing problems with CMake with workarounds suggested by people 10+ years ago that are still valid today.  So relying on web searches is a double-edge sword.

How can I state this in a compact way?

Author: I shortened this some in https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/af525be727200d27e6db242b72b4c1fb1ffe61b3

Reviewer: > So relying on web searches is a double-edge sword.

I think I see what you mean. But, I feel like this is true (to one extent or another) about almost *any* HPC software. I don't see what is *unique* about CMake in this regard that requires it being explicitly called out or explicitly alerting readers to the possiblity. Welcome to tech.

Author: That is true.  For example, advice on C++ given 10 years ago related to C++98 may not be the best advance with C++17.  So I guess it goes without saying as languages and tools evolve, what was good advice X years ago is not the best advice today.

So with that I can take out all mentions of web searches.

Author: Hopefully mostly addressed in commit https://github.com/betterscientificsoftware/bssw.io/pull/1276/commits/81408104691cb437122181e72f517b02feffe09a
